File: backend/routers/bookmarks.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from ..database import get_db
from ..models import Bookmark
from ..schemas import BookmarkCreate, BookmarkUpdate, BookmarkResponse, BookmarkReorderItem

router = APIRouter()


@router.get("", response_model=list[BookmarkResponse])
def list_bookmarks(db: Session = Depends(get_db)):
    return db.query(Bookmark).order_by(Bookmark.sort_order, Bookmark.created_at).all()


@router.post("", response_model=BookmarkResponse)
def create_bookmark(body: BookmarkCreate, db: Session = Depends(get_db)):
    bm = Bookmark(**body.model_dump())
    db.add(bm)
    db.commit()
    db.refresh(bm)
    return bm


@router.post("/reorder")
def reorder_bookmarks(body: list[BookmarkReorderItem], db: Session = Depends(get_db)):
    for item in body:
        bm = db.get(Bookmark, item.id)
        if bm:
            bm.sort_order = item.sort_order
    db.commit()
    return {"ok": True}


@router.put("/{bookmark_id}", response_model=BookmarkResponse)
def update_bookmark(bookmark_id: str, body: BookmarkUpdate, db: Session = Depends(get_db)):
    bm = db.get(Bookmark, bookmark_id)
    if not bm:
        raise HTTPException(404, "Bookmark not found")
    for k, v in body.model_dump().items():
        setattr(bm, k, v)
    db.commit()
    db.refresh(bm)
    return bm


@router.delete("/{bookmark_id}")
def delete_bookmark(bookmark_id: str, db: Session = Depends(get_db)):
    bm = db.get(Bookmark, bookmark_id)
    if not bm:
        raise HTTPException(404, "Bookmark not found")
    db.delete(bm)
    db.commit()
    return {"ok": True}

File: frontend/src/api/bookmarks.ts
import { apiFetch } from "./client";
import type { Bookmark } from "../types";

export const getBookmarks = () => apiFetch<Bookmark[]>("/bookmarks");
export const createBookmark = (data: Omit<Bookmark, "id" | "created_at">) =>
  apiFetch<Bookmark>("/bookmarks", { method: "POST", body: JSON.stringify(data) });
export const updateBookmark = (id: string, data: Omit<Bookmark, "id" | "created_at">) =>
  apiFetch<Bookmark>(`/bookmarks/${id}`, { method: "PUT", body: JSON.stringify(data) });
export const deleteBookmark = (id: string) =>
  apiFetch<{ ok: boolean }>(`/bookmarks/${id}`, { method: "DELETE" });
export const reorderBookmarks = (items: { id: string; sort_order: number }[]) =>
  apiFetch<{ ok: boolean }>("/bookmarks/reorder", { method: "POST", body: JSON.stringify(items) });
